File: Java/Sorting/Selection/Selection.java
import java.util.ArrayList;

public class Selection {

   static java.util.ArrayList<Integer> selctionSort(java.util.ArrayList<Integer> array) {
      int currentIteration = 0;
      int minimum = 100;
      while (currentIteration < array.size()) {
         
         for (int i = currentIteration; i < array.size(); i++) {
            if (minimum > array.get(i)) {
               minimum = array.get(i);
            }
         }
         int indexOfMinimum = array.indexOf(minimum);
         int temp = array.indexOf(currentIteration);
         System.out.println(minimum);
         array.set(currentIteration, minimum);
         array.set(indexOfMinimum, temp);
         minimum = 100;
         currentIteration++;
      }
      return array;
   }

   public static void main(String[] str) {
      java.util.ArrayList<Integer> arr = new java.util.ArrayList<>();
      arr.add(4);
      arr.add(3);
      arr.add(2);
      arr.add(1);
      var result = selctionSort(arr);
      for(int i = 0 ; i < result.size() ; i++){
         System.out.println(result.get(i));
      }

   }
}

File: Javascript/Algoritms/Sorting Algorithms/Selection Sort/Selection.js
// Selection Sorting is based on select the minumum and put it in the front of an array !
// ? this is naive implementation made by me
// the time complexity of this algotithms is o(n^2) ! bad
function selectionSort(arr) {
  let currentIteration = 0;
  let minimum = Infinity;
  while (currentIteration < arr.length) {
    for (let i = currentIteration; i < arr.length; i++) {
      if (arr[i] < minimum) {
        minimum = arr[i];
      }
    }
    let indexOfMinimum = arr.lastIndexOf(minimum);
    [arr[indexOfMinimum], arr[currentIteration]] = [
      arr[currentIteration],
      arr[indexOfMinimum],
    ];
    minimum = Infinity;
    currentIteration++;
  }
  return arr;
}

// let result = selectionSort([1, 3, 2, 0.5]);
// console.log(result);

function betterSelectionSort(arr) {
  for (let i = 0; i < arr.length; i++) {
    let lowest = i; // lowest = 0 ;
    for (let j = i + 1; j < arr.length; j++) { // j = 1 => arr.length
      if (arr[j] < arr[lowest]) { // 1 < 3 true 
        lowest = j; // lowest = 1;
      }
    }
    if (i !== lowest) [arr[i], arr[lowest]] = [arr[lowest], arr[i]];
  }
  return arr;
}

const result = betterSelectionSort([3, 1 , 1, 2, 2 , 2 , 0.5]);
console.log(result);
